content/blog/2025-12-16-loop-parallelization-proposal.md.md

```diff
@@ -0,0 +1,552 @@
+- **Element-wise**: Simple element-wise computations
+
+### 2. PolyBench/C Suite
+30 real-world scientific computing benchmarks including:
+- Linear algebra (gemm, syrk, seidel-2d)
+- Stencil computations (jacobi, fdtd-2d)
+- Data mining (correlation, covariance)
+
+## Methodology
+
+For each implementation, we:
+- Compiled benchmarks with `-O2` optimization
+- Ran with thread counts: 2, 4, 8
+- Compared against **serial baseline** (original code compiled without any parallelization pass applied)
+- Ran 5-10 iterations and computed mean/std deviation
+- Measured execution time using benchmark-internal timers
+
+**Hardware:** Apple M3 (8-10 cores)
+**LLVM Version:** 18.1.8
+**Compiler Flags:** `-O2 -fopenmp`
+
+# Results and Analysis
+
+## Test Benchmark Results
+
+![Test Performance Comparison](figures/tests_performance_comparison.png)
+
+The synthetic benchmarks reveal surprising results (speedup relative to serial baseline):
+
+| Implementation | Array Ops (8T) | Scale Offset (8T) | Elementwise (8T) |
+|----------------|----------------|-------------------|------------------|
+| Serial Baseline| 1.00x          | 1.00x             | 1.00x            |
+| Original       | 0.98x          | 0.99x             | 1.01x            |
+| Fusion         | 0.96x          | 0.97x             | 0.98x            |
+| Fusion+Shared  | 0.95x          | 0.96x             | 0.99x            |
+
+**Key Observations:**
+1. **Minimal speedup across all implementations** (< 1.05x)
+2. **Fusion variants actually slower** than the original
+3. **Performance degrades with optimization complexity**
+
+## PolyBench Results
+
+![PolyBench Speedup Comparison](figures/polybench_speedup_comparison.png)
+
+### Parallelization Coverage
+
+Out of 30 PolyBench benchmarks:
+- **Original**: 6 benchmarks parallelized (20%)
+- **Fusion**: 6 benchmarks parallelized (20%)
+- **Fusion+Shared**: 6 benchmarks parallelized (20%)
+
+All three implementations identified the same set of parallelizable loops, confirming our conservative analysis is consistent.
+
+### Speedup Analysis
+
+![PolyBench Speedup Scaling](figures/polybench_speedup_scaling.png)
+
+Average speedup on parallelizable benchmarks (8 threads) relative to serial baseline:
+
+| Implementation | Mean Speedup | Median | Best Case |
+|----------------|--------------|--------|-----------|
+| Serial Baseline| 1.00x        | 1.00x  | 1.00x     |
+| Original       | 1.12x        | 1.08x  | 1.85x     |
+| Fusion         | 1.09x        | 1.05x  | 1.78x     |
+| Fusion+Shared  | 1.08x        | 1.04x  | 1.76x     |
+
+**Best Performing Benchmarks (Original, 8T):**
+- `nussinov`: 1.85x speedup
+- `gesummv`: 1.34x speedup
+- `atax`: 1.28x speedup
+
+# Why Fusion Didn't Help: Deep Dive
+
+Our initial hypothesis—that loop fusion would improve performance—was contradicted by the data. Here's our analysis of why:
+
+## 1. Memory Bandwidth Bottleneck
+
+The synthetic benchmarks are **memory-bound**, not compute-bound:
```

Comment:
Hmm, I thought this section was about the PolyBench benchmarks? I don't think you gave any fusion performance results for the hand-rolled kernels?